### ISSUE TITLE
Load random backgrounds (BG) for different rooms

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#79]
## Feature Description

Each event should have its own BG randomly-ish loaded when player enters a new room.

## Motivation

We don't have a lot of BG arts for now, but it's still good to implement the feature for later.

## Acceptance Criteria

- Combat rooms randomly select a BG from all possible BG for combats. This selection is done so that each BG appears once in a rotation (so that if you have BG A, B, C; you see each once and not B C C A for example)
- Dialogue rooms, shops and heals should have their own BG

## Proposed Solution

Each dialogue event has a linked BG, same for heal and shop.
For combats, we generate a shuffled list of all the possible BG. When entering a combat room, we pick the first (or last, or the middle idk but something simple) element, and we remove it from the list. When the list is empty, create a new one. Maybe just ensure that the last seen BG is not the same as the first BG of the new list.

I also think that we should have a classification of BG depending on where we can find them. We have 3 parts of the tree, so we should have least have a way to say that we want the elements of roots only for the root part (first map), the trunk and the canopy as separated. We might later on also want to have separation within the same map, for example some BG only for the first part of the roots, then some other BG for the second part.
My idea for that is to have a dictionary with tags and a list of possible BG. 
Like so:
```GDScript
{
"1.1" : ["BG_name_A", "BG_name_B" ...],
"1.2" : [ bla bla],
"2.1" : [more stuff],
"3.1" : [even more BG]
}
```

With the tag "X.Y" with X the number of the map (1 for the roots, 2 the trunk and 3 the canopy). The Y is used as a sub selection, for example if we want upper a lower part on the roots for the BG, we have the sub selection 1 and 2 (and we can add more if we want).
I also think that we should have a dictionary of link for the BG. That way if a BG is both is 1.1 and 1.2 for example, and we change the resource link, we only change it in one place:
```
{
"BG_name_A": link/to/BG_A,
"BG_name_B": link/to/BG_B
...
}
```

Using dictionaries like that instead of just putting the BG into dedicated folders allows to not have the same image in two places. But we could decide to just have repetitions and just separate into folders, as this will probably be much less heavy to write (just put stuff into the sub_folder where you want the BG to appear, and no need to add its name into the code).
If we go for this option, we just need a dictionary with the sub sections and the associated folder name, tag X.Y as the first dictionary example.

After writing it, I think I might actually like option 2 more

## Additional Context

This probably makes more sense if it's done after #57 and #74

**Comment by BishopJohnson** [comment]
I'm thinking we could try using enums as flags to decide which kind of events and locations in the tree that the backgrounds may be used in (may also be used for mobs or other stuff too in the future). It may require a little more logic handling though for the flags, since it seems that there was a regression with being able to use enums as flags going from Godot 3 to Godot 4 (see: https://github.com/godotengine/godot/issues/74418), but I think we could work around this for the time being.

There is one thing I was wondering about however. In the first option @Turtyo proposed, they mentioned sub-selection for the regions of the tree. Is this a for sure feature for the game and are the range of sub-regions fixed or could it be an arbitrarily large value? Because I'm thinking if there are supposed to be sub-regions that are more special (e.g. roots beginning/roots middle/roots end), then we could treat those as their own regions.

**Comment by BishopJohnson** [comment]
I'm thinking we could handle deciding which backgrounds are allowed by using enum flags to denote different regions and events (could also use it for things like deciding allowed mob spawns later if we want to). However, it seems that there may have been a regression going from Godot 3 to Godot 4 for using enum types themselves as flags in gdscript, but I think we can work around this for the time being.

Also one thing I was wondering. In the first option, @Turtyo mentioned having sub-selections for the different regions. I was wondering if this is a definite feature for the game and if these sub-selections have a fixed range or if they can be arbitrarily large? If they are a thing, but are fairly constrained, then I think we could maybe treat them as their own regions (e.g. roots begin/roots middle/roots end).

**Comment by Turtyo** [comment]
I'm not sure how flags would work, since we would still need something to check all the BG that have a certain tag.
For the sub-selections, I think we can with fixed range first ?

I'm giving a solution I thought of. You make a centralized class that gets the resources for you
Let's say we choose to make 3 sections for the roots, so roots-1, roots-2, roots-3
and say we have 10 floors, BG for the section 1 are between floors 1-3, then 4-6, then 7-10
The loading class only holds a list of when sections start (for example it could be the list `[1, 4, 7]`). When we get to a new floor the class checks if it needs to load new BG from a new directory.
Select the relevant directory (like roots-2 for example) and load the list of BG (don't load the image, just the names, we could say the names are just numbers from 1-7 for example). Apply the shuffle that I described that I described at the start of my previous proposed solution.

Now, we could say it could take too long when loading a new scene to load all the possibles BG. So here is what we do. Instead of having the list `[1, 4, 7]` we make the list `[1, 3, 6]`. We make a variable `next_BG_to_load` which is always ready to give a BG to a scene that needs it. The list of all possibles BG is called `loadable_BG`
When we get to a new floor:
- scene requests the `next_BG_to_load`, you give it (you give the reference to the file actually, and the scene handles loading it)
- check if we are at a switch floor. If we are, load the names of the new BG to be loaded in `loadable_BG` (in the background on a non blocking thread, so it doesn't matter if it's a bit longer than a frame, because the next time we need a BG is when the player finishes the next event, and we will clearly be done loading by then)
- select a new BG from `loadable_BG` after having shuffled it (like the last element of the list for example) to be the new `next_BG_to_load`

That works since when we get to floor 3, we chose that BG on floor 2. So we get a BG of the correct section on floor 3. Then we are at a switch floor, so we load the BG of the section 4-6, and we select a BG for that section. When we load on the section 4-6, we get a BG of that section.
If you want to add have a certain BG on a certain range of floor, you just need to put the image in the correct folder.

**Comment by BishopJohnson** [comment]
I think using directories to sort the backgrounds would lead to having duplicate images if we wanted to use the same image across multiple floor groups or in cases if there was just one image, then it would have to be duplicated across those directories to prevent problems from when there's no image to be loaded for the floor.

In my opinion, it may be better to have a class that handles loading and caching resources that contain backgrounds and can be called upon by any class that may need one. As for ensuring that backgrounds aren't used consecutively, we could either create a separate class that would be solely responsible for that functionality or incorporate it into an existing class that would reasonably be responsible and who's data would persist between scenes.

Also, I think we should try to use enums or some other constants/variables to categorize which background may be used on which type of floor. Mainly to avoid using magic numbers, but as I mentioned before if the enums were usable as flags, then we could allow certain backgrounds to be usable on multiple floors if we want to.

**Comment by Turtyo** [comment]
You are not wrong about having to duplicate images, there might be problems of sync later if we don't think of changing the same image across different directories.
I think the class that handles loading the BG can also be the one ensuring no consecutive use of the floors (less class will make it clearer, and it already has all the data it needs to do so).

How would you make this loading class *exactly* ? For flags the doc seems to say we can [do it like that](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_exports.html#exporting-bit-flags). Even if it's not enums, it should work pretty well I think

**Comment by Turtyo** [comment]
I'm starting to wonder if we even want that sub-selection. Because it would probably be much easier to do if we didn't have that, and it's not even a thing that was given to the artists team so basically we can only cut by tree section (eg roots, trunk, canopy)

**Comment by BishopJohnson** [comment]
For the enums, I was thinking of working around the regression in godot 4 by using an array of enums and using their integer values to create and cache a bitmask that is the flags for the background. This way, if there's a need to do something similar for other elements, like mobs, then the enums themselves may be reused for them too. For the loading class itself, I was think it could be a singleton that either stores the bg resources in an array or it can fetch them from a single resource directory and then has them cache their bitmasks.

As for subsections, if they weren't originally planned, then that's probably something that ought to be decided on sooner rather than later. I could imagine maybe wanting to have subsections for maybe a boss level, but that could also just as easily be an event separate from a normal mob event.

**Comment by Turtyo** [comment]
Boss level would be specifically made so there probably wouldn't be a need for this sub-selection.
It could make sense to have sub-selection for mobs though, as we could imagine having weaker mobs at the start of a floor than at its end

Okay for the bitmask, it seems fine to me. If we do a sub-selection, the bitmask can do it I think ?

**Comment by Turtyo** [comment]
Okay so since we will probably want sub-selection for the mobs anyway, I think it's a good think if we already have it for maps (even if we don't use it for BG, it will be used for mobs)

**Comment by BishopJohnson** [comment]
Got it. Also to clarify, is the plan just three sub-selections for the different areas of the tree? (e.g. "roots early", "roots middle", and "roots end")

**Comment by Turtyo** [comment]
I think so yes. Does it matter ?

**Comment by BishopJohnson** [comment]
Just want to make sure.

**Comment by BishopJohnson** [comment]
For the background nodes in the scenes, I was thinking we could use the `ParallaxBackground` node with a child texture node, as it seems to consistently have the background texture behind everything else in the scene.

**Comment by Turtyo** [comment]
Makes sense to me ! If you don't have anymore questions, I think you can go for it. Don't forget to confirm the issue by adding the label

**Comment by BishopJohnson** [comment]
I got one more question regarding the scope of implementing this feature. I noticed that some of the events are still a WIP. Would we still want to implement the backgrounds for those WIP events or would it be alright to implement it just for the mob and dialogue event?

**Comment by Turtyo** [comment]
Even if they are still WIP, we probably won't touch a lot of the BG after it's done. But those events will also not load their BG in the same way the map does, as they might have a BG specific to them (meaning we choose the BG instead of having a random selection).

I think in a first time, you can leave the WIP on the side, we can see about those later.

**Comment by BishopJohnson** [comment]
@Turtyo, I'm looking for some clarification regarding the dialogue, heal, and shop backgrounds. When the acceptance criteria states that these events ought to have their own backgrounds, does that mean that for example all dialogue events may share the same background with each other or does it mean that each individual dialogue event may have its own background? And if it is the second option, then would each event then also need a background for each tree area in case it gets generated some place different?

**Comment by Turtyo** [comment]
Each individual dialogue event may have its own background.
I'm not exactly sure to understand the second question, as I thought that the idea of having a bitmask was to allow events to have a different background depending on the tree area ?
